src/smith/differentiable_numerics/tests/test_explicit_dynamics.cpp

```diff
@@ -0,0 +1,420 @@
+
+  double K;  ///< bulk modulus
+  double G;  ///< shear modulus
+  double density0;
+};
+
+struct MeshFixture : public testing::Test {
+  static constexpr int dim = 2;
+  static constexpr int disp_order = 1;
+
+  static constexpr int scalar_field_order = 1;
+  using DensitySpace = smith::L2<scalar_field_order>;
+  using VectorSpace = smith::H1<disp_order, dim>;
+
+  using SolidMaterial = NeoHookeanWithFixedDensity;
+
+  static constexpr double gravity = -9.0;
+
+  enum STATE
+  {
+    DISP,
+    VELO,
+    ACCEL
+  };
+
+  enum PARAMS
+  {
+    DENSITY
+  };
+
+  enum FIELD
+  {
+    F_DISP,
+    F_VELO,
+    F_ACCEL,
+    F_DENSITY
+  };
+
+  void SetUp()
+  {
+    MPI_Barrier(MPI_COMM_WORLD);
+    smith::StateManager::initialize(datastore_, "solid_dynamics");
+
+    // create mesh
+    auto mfem_shape = mfem::Element::QUADRILATERAL;  // mfem::Element::TRIANGLE;
+    double length = 0.5;
+    double width = 2.0;
+    mesh = std::make_shared<smith::Mesh>(mfem::Mesh::MakeCartesian2D(5, 5, mfem_shape, true, length, width), MESHTAG, 0,
+                                         0);
+    // checkpointing graph
+    checkpointer_ = std::make_shared<gretl::DataStore>(200);
+
+    // create residual evaluator
+    const double density = 1.0;
+    std::string physics_name = "solid";
+
+    shape_disp = std::make_unique<smith::FieldState>(
+        create_field_state(*checkpointer_, VectorSpace{}, physics_name + "_shape_displacement", mesh->tag()));
+    auto disp = create_field_state(*checkpointer_, VectorSpace{}, physics_name + "_displacement", mesh->tag());
+    auto velo = create_field_state(*checkpointer_, VectorSpace{}, physics_name + "_velocity", mesh->tag());
+    auto accel = create_field_state(*checkpointer_, VectorSpace{}, physics_name + "_acceleration", mesh->tag());
+    auto density0 = create_field_state(*checkpointer_, DensitySpace{}, physics_name + "_density", mesh->tag());
+
+    *disp.get() = 0.0;
+    *velo.get() = 0.0;
+    *accel.get() = 0.0;
+    *density0.get() = density;
+
+    initial_states = {disp, velo, accel};
+    params = {density0};
+    std::vector<smith::FieldState> states{disp, velo, accel};
+
+    auto solid_mechanics_residual = smith::create_solid_weak_form<disp_order, dim, DensitySpace>(
+        physics_name, mesh, getConstFieldPointers(states), getConstFieldPointers(params));
+
+    SolidMaterial mat;
+    mat.density0 = density;
+    mat.K = 1.0;
+    mat.G = 0.5;
+
+    solid_mechanics_residual->setMaterial(smith::DependsOn<>{}, mesh->entireBodyName(), mat);
+
+    solid_mechanics_residual->addBodySource(mesh->entireBodyName(), [](auto /*time*/, auto X) {
+      auto b = 0.0 * X;
+      b[1] = gravity;
+      return b;
+    });
+
+    // create mass evaluator and state in order to be able to create a diagonalized mass matrix
+    std::string mass_residual_name = "mass";
+    auto solid_mass_residual = smith::createSolidMassWeakForm<VectorSpace::components, VectorSpace, DensitySpace>(
```

Comment:
Looks like the dFEM version would drop right in here. Did you get a chance to try this out?